bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4362,6 +4478,628 @@ void ReplicaImp::setConflictDetectionBlockId(const ClientRequestMsg &clientReqMs
+
+    if (config_.timeServiceEnabled && (req.flags() & MsgFlag::TIME_SERVICE_FLAG) && reqIdx == 0) {
+      numOfSpecialReqs++;
+      reqIdx++;
+      continue;
+    }
+
+    // TODO(GG): should be verified in the validation of the PrePrepare . Consdier to remove this assert
+    ConcordAssert(!(req.flags() & MsgFlag::TIME_SERVICE_FLAG));
+    if ((req.flags() & MsgFlag::KEY_EXCHANGE_FLAG) || (req.flags() & MsgFlag::RECONFIG_FLAG)) {
+      numOfSpecialReqs++;
+      reqIdx++;
+      continue;
+    }
+
+    if (req.requestLength() == 0) {
+      if (clientId == currentPrimary())
+        ++numValidNoOps;  // TODO(GG): do we want empty requests from non-primary replicas?
+      reqIdx++;
+      continue;
+    }
+
+    const bool validClient = isValidClient(clientId);
+    if (!validClient) {
+      ++numInvalidClients;
+      //          LOG_WARN(GL, "The client is not valid" << KVLOG(clientId));
+      reqIdx++;
+      continue;
+    }
+
+    if (isReplyAlreadySentToClient(clientId, req.requestSeqNum())) {
+      auto replyMsg = clientsManager->allocateReplyFromSavedOne(clientId, req.requestSeqNum(), currentPrimary());
+      if (replyMsg) {
+        send(replyMsg.get(), clientId);
+      }
+      reqIdx++;
+      continue;
+    }
+    if (allowParallelExecution && !shouldRunRequestsInParallel) shouldRunRequestsInParallel = true;
+    requestSet.set(reqIdx);
+    reqIdx++;
+  }
+  reqIter.restart();
+
+  if (ps_) {
+    DescriptorOfLastExecution execDesc{lastExecutedSeqNum + 1, requestSet};
+    ps_->beginWriteTran();
+    ps_->setDescriptorOfLastExecution(execDesc);
+    ps_->endWriteTran();
+  }
+}
+
+void ReplicaImp::executeAllPreprepareRequests(bool allowParallelExecution,
+                                              bool shouldRunRequestsInParallel,
+                                              uint16_t numOfSpecialReqs,
+                                              PrePrepareMsg *ppMsg,
+                                              Bitmap &requestSet,
+                                              bool recoverFromErrorInRequestsExecution) {
+  // !allowParallelExecution --> !shouldRunRequestsInParallel
+  ConcordAssert(allowParallelExecution || !shouldRunRequestsInParallel);
+
+  //////////////////////////////////////////////////////////////////////
+  // Phase 2: execute requests + send replies
+  // In this phase the application state may be changed. We also change data in the state transfer module.
+  // TODO(GG): Explain what happens in recovery mode (what are the requirements from  the application, and from the
+  // state transfer module.
+  //////////////////////////////////////////////////////////////////////
+
+  Timestamp time;
+
+  if (numOfSpecialReqs > 0) executeSpecialRequests(ppMsg, numOfSpecialReqs, recoverFromErrorInRequestsExecution, time);
+
+  ConcordAssert(activeExecutions_ == 0);
+  activeExecutions_ = 1;
+  if (shouldRunRequestsInParallel) {
+    PostExecJob *j = new PostExecJob(ppMsg, requestSet, time, *this);
```

Comment:
validate that threadPool request execution is deterministic if we have one thread

Reply:
It is.